implementations/rust/ockam/ockam_api/src/cloud/project.rs

```diff
@@ -296,17 +290,25 @@ mod node {
             .await
         }
 
-        pub(crate) async fn list_projects(
-            &mut self,
-            ctx: &mut Context,
-            dec: &mut Decoder<'_>,
+        pub async fn list_projects(
+            &self,
+            ctx: &Context,
+            route: &MultiAddr,
+        ) -> Result<Vec<Project>> {
+            let bytes = self
+                .list_projects_response(ctx, CloudRequestWrapper::bare(route))
+                .await?;
+            Response::parse_response_body(bytes.as_slice())
+        }
+
+        pub(crate) async fn list_projects_response(
```

Comment:
I think this will come in very handy. I'm not sure I like the `_response` suffix though. Perhaps, in another PR, we could introduce a new type that can handle both the typed and encoded response so we can write something like `self.list_projects().encode()` and avoid having to write two methods.

Reply:
I am also not a big fan of that name but I don't see how we can avoid 2 methods. 

One is public and returns a decoded value and the other one is crate local and keeps the undecoded response.
For the first one it is better to have a direct call: `list_projects()`. For the second one it is better to avoid the unnecessary decode + encode that would happen with `list_projects().encode()`. What kind of data type are you thinking of to fix this?

Reply:
I was playing around with this idea and I realized we would have do it the other way around: `list_projects()` returns the encoded body, and then `list_projects().decode()` would return the typed struct. 

As you pointed in the description, we need to return well-typed structs from the API so we can stop moving the responsibility of decoding the body to its expected type up to the API callers (cli, app), which is very uncomfortable.

I was thinking something along these lines:

```rust
struct ResponseBody<T>
where
    T: for<'b> Decode<'b, ()>,
{
    inner: Vec<u8>,
    t: PhantomData<T>,
}

impl<T> ResponseBody<T>
where
    T: for<'b> Decode<'b, ()>,
{
    pub fn new(encoded: Vec<u8>) -> Self {
        Self {
            inner: encoded,
            t: PhantomData::default(),
        }
    }

    pub fn inner(&self) -> &[u8] {
        &self.inner
    }

    pub fn decode(&self) -> Result<T> {
        Response::parse_response_body(&self.inner)
    }
}

#[derive(Debug, Clone, Encode, Decode)]
#[cbor(map)]
struct Space {
    #[n(0)]
    name: String,
}

fn create_space() -> Result<ResponseBody<Space>> {
    let raw_body: Vec<u8> = {
        // handle request
        vec![]
    };
    Ok(ResponseBody::new(raw_body))
}

fn handler() -> Result<()> {
    let body = create_space()?;
    let space = body.decode()?;
    Ok(())
}
```